Разработка нового функционала для Fiilmorate

Alex-Galvinauskas opened the pull request.

Alex-Galvinauskas committed:
refactor: Дописать в schema.sql таблицу режиссеров

Дописал таблицу режиссеров CREATE TABLE IF NOT EXISTS directors
Дописал связь фильмов и режиссеров многие-ко-многим CREATE TABLE IF NOT EXISTS directors
Добавил индексы для производительности

Alex-Galvinauskas committed:
feat: Добавить модель и DTO для режиссеров

Alex-Galvinauskas committed:
refactor: Обновить классы фильмов для работы с режиссерами

Добавить режиссеров в модель и DTO Film;
Добавить контракт для MapStruct DirectorMapper
Добавить режиссеров в FilmMapper для DTO ↔ Entity

Alex-Galvinauskas committed:
feat: создать слой БД для режиссеров

Alex-Galvinauskas committed:
feat: реализована функциональность режиссеров с полным CRUD и интеграцией с фильмами

- DirectorDbStorage с CRUD операциями для режиссеров
- DirectorService и DirectorServiceImpl для бизнес-логики
- FilmDbStorage: добавлена работа с режиссерами при создании/обновлении фильмов
- FilmServiceImpl: добавлена валидация режиссеров и новый метод getFilmsByDirector
- Добавлена сортировка фильмов режиссера по году или количеству лайков
- Оптимизированы импорты

Alex-Galvinauskas committed:
feat: добавлен DirectorController и эндпоинт для фильмов режиссера

- Добавлен DirectorController с CRUD операциями для режиссеров
- В FilmController добавлен эндпоинт GET /films/director/{directorId}
- Поддержка сортировки фильмов режиссера по лайкам и году выхода

Alex-Galvinauskas committed:
refactor: поправил названия методов чтобы не путаться

forsidenis committed:
Функциональность «Популярные фильмы». 2 SP, выполнено

forsidenis committed:
Функциональность «Популярные фильмы». 2 SP, выполнено

forsidenis committed:
Merge remote-tracking branch 'origin/add-most-populars' into add-most-populars

forsidenis committed:
Функциональность «Популярные фильмы». 2 SP, выполнено

forsidenis committed:
Убрал логгер

forsidenis committed:
Убрал логгер, и поправил на javadoc

Alex-Galvinauskas committed:
test: написать тесты для controller и DBstorage режиссеров

Alex-Galvinauskas committed:
fix: исправить работу валидации фильмов

fix: buildDirectorFilmsQuery

checkstyle: убрать неиспользуемый импорт

fix: исправление работы эндпоинта getFilmsByDirector

fix: заменить валидацию имени режиссера

fix: исправлен SQL запрос для получения фильмов режиссера

Исправлена сортировка по лайкам в методе getFilmsByDirector
Добавлена проверка существования режиссера
Исправлена группировка в SQL запросе

Revert "fix: исправление имени JSON поля для режиссеров в FilmDTO"

This reverts commit 56eaeffe19cd46c2b4fd1d2492cac567884ad9f8.

fix: исправление имени JSON поля для режиссеров в FilmDTO

Использовано @JsonProperty("director") для соответствия API спецификации.

Alex-Galvinauskas committed:
fix: исправление имени JSON поля для режиссеров в FilmDTO

Использовано @JsonProperty("director") для соответствия API спецификации.

Alex-Galvinauskas committed:
Revert "fix: исправление имени JSON поля для режиссеров в FilmDTO"

This reverts commit 56eaeffe19cd46c2b4fd1d2492cac567884ad9f8.

Alex-Galvinauskas committed:
fix: исправлен SQL запрос для получения фильмов режиссера

Исправлена сортировка по лайкам в методе getFilmsByDirector
Добавлена проверка существования режиссера
Исправлена группировка в SQL запросе

Alex-Galvinauskas committed:
fix: заменить валидацию имени режиссера

Alex-Galvinauskas committed:
fix: исправление работы эндпоинта getFilmsByDirector

Alex-Galvinauskas committed:
checkstyle: убрать неиспользуемый импорт

Alex-Galvinauskas committed:
fix: buildDirectorFilmsQuery

Alex-Galvinauskas committed:
fix: исправить работу валидации фильмов

Stepanovmaxvl committed:
Реализована функциональность Отзывы

Stepanovmaxvl committed:
Корректировка метода класса ReviewDbStorage после неуспешных QG.Реализована функциональность Отзывы

Stepanovmaxvl committed:
Корректировка метода класса ReviewServiceImpl после неуспешных QG.Реализована функциональность Отзывы

Stepanovmaxvl committed:
add-reviews

forsidenis committed:
Функциональность «Рекомендации». 3 SP. Выполнил

forsidenis committed:
удалил неиспользуемый импорт FilmDTO

forsidenis committed:
удалил неиспользуемый импорты

forsidenis committed:
еще раз удалил неиспользуемые импорты

forsidenis committed:
поправил для прохождения postman

Stepanovmaxvl committed:
Add add-reviews

Alex-Galvinauskas committed:
fix: исправить валидацию фильмов с учетом режиссеров

Alex-Galvinauskas committed:
test: исправить работу тестов

Alex-Galvinauskas committed:
Merge branch 'add-director' of github.com:Alex-Galvinauskas/java-filmorate into add-director

Alex-Galvinauskas committed:
fix: изменить имя класса тестов
test: исправить

Alex-Galvinauskas committed:
fix: изменить имя класса тестов

Alex-Galvinauskas committed:
Merge branch 'add-director' of github.com:Alex-Galvinauskas/java-filmorate into add-director

Alex-Galvinauskas committed:
fix: исправить N + 1 в getAllFilms, добавить @Transactional в менеджеры бд

Alex-Galvinauskas committed:
Merge pull request #8 from forsidenis/add-most-populars

add-most-populars

Alex-Galvinauskas committed:
Merge branch 'develop' into add-recommendations

Alex-Galvinauskas committed:
Update RecommendationServiceImplTest.java

Alex-Galvinauskas committed:
Merge branch 'develop' into add-director

Alex-Galvinauskas committed:
Merge pull request #9 from Alex-Galvinauskas/add-director

Add director

forsidenis committed:
поправил замечания

Alex-Galvinauskas committed:
Revert "Update RecommendationServiceImplTest.java"

This reverts commit 8864d5b2c41d57881e7aafb20a0d53ed2847a4df.

Alex-Galvinauskas committed:
Revert "Merge branch 'develop' into add-recommendations"

This reverts commit c3f04a94371def8856e62e5327b890c5e383301a, reversing
changes made to 119c0438548cc8f06d6ad2107aad7b07daf9dd63.

Alex-Galvinauskas committed:
Merge branch 'develop' into add-recommendations

Alex-Galvinauskas committed:
Update FilmDbStorage.java

Alex-Galvinauskas committed:
Update FilmDbStorage.java

Alex-Galvinauskas committed:
Update FilmDbStorage.java

Alex-Galvinauskas committed:
Update RecommendationServiceImplTest.java

Alex-Galvinauskas committed:
Update RecommendationServiceImplTest.java

Alex-Galvinauskas committed:
Update RecommendationServiceImplTest.java

Alex-Galvinauskas committed:
Update RecommendationIntegrationTest.java

Alex-Galvinauskas committed:
Update RecommendationIntegrationTest.java

Alex-Galvinauskas committed:
Update RecommendationIntegrationTest.java

Alex-Galvinauskas committed:
Merge pull request #11 from forsidenis/add-recommendations

Функциональность «Рекомендации». 3 SP. Выполнил

Alex-Galvinauskas committed:
feat: реализация ленты событий для пользователей

P.S требуется доработка (нужна реализация отзывов)

Alex-Galvinauskas committed:
Update FeedController.java

Stepanovmaxvl committed:
Сделаны правки после ревью

Stepanovmaxvl committed:
Merge branch 'develop' into add-reviews

Stepanovmaxvl committed:
Сделаны правки после ревью и чек-стайла

Stepanovmaxvl committed:
Сделаны правки

Stepanovmaxvl committed:
Сделаны правки

Stepanovmaxvl committed:
Сделаны правки

Commit:
Длбавлена Функциональность «Удаление фильмов и пользователей»

Maxim0911 committed:
Merge branch 'develop' into add-remove-endpoint

Commit:
Добавлена Функциональность «Удаление фильмов и пользователей»

Commit:
Добавлена Функциональность «Удаление фильмов и пользователей»

Alex-Galvinauskas committed:
xheckstyle: убрать пробелы

Alex-Galvinauskas committed:
fix: исправить shema.sql

Alex-Galvinauskas committed:
Merge pull request #10 from Alex-Galvinauskas/add-reviews

Реализована функциональность Отзывы

Alex-Galvinauskas committed:
Merge branch 'add-feed1' into add-feed

# Conflicts:
#	src/main/resources/schema.sql

Alex-Galvinauskas committed:
refactor: добавить отзывы в ленту событий

Alex-Galvinauskas committed:
refactor: исправить запись событий дружбы в ленту для обоих пользователей

Теперь при добавлении друга:
- В ленту пользователя A: userId=A, actorId=A, entityId=B
- В ленту пользователя B: userId=B, actorId=B, entityId=A

При удалении друга:
- В ленту пользователя A: userId=A, actorId=A, entityId=B
- В ленту пользователя B: userId=B, actorId=B, entityId=A

Оба пользователя теперь корректно видят события дружбы в своих лентах активности.

twiksokyan committed:
add search by film title and director name part

twiks-yp-okyan committed:
fix unused imports & add postman tests to gitignore

Alex-Galvinauskas committed:
refactor: убрать ненужные методы

вынести дублирующийся код записи событий в отдельные методы.

Alex-Galvinauskas committed:
Revert "Добавлена Функциональность «Удаление фильмов и пользователей»"

This reverts commit 96edadd9d3f9991cf9981b9d45b4c3e2ba0cb5df.

Alex-Galvinauskas committed:
style: убрать завершающий пробел

Alex-Galvinauskas committed:
style: добавить закрывающую скобку

Alex-Galvinauskas committed:
style: довести до соответствия chekstyle

Alex-Galvinauskas committed:
Merge pull request #18 from Alex-Galvinauskas/add-remove-endpoint

Добавлена Функциональность «Удаление фильмов и пользователей»

Alex-Galvinauskas committed:
Merge branch 'develop' into add-feed

Alex-Galvinauskas committed:
fix: удалить ненужную проверку пользователя

Alex-Galvinauskas committed:
Merge pull request #21 from Alex-Galvinauskas/add-feed

Добавлена функциональность ленты событий

Commit:
Добавлена Функциональность «Общие фильмы»

twiks-yp-okyan committed:
fix error in SQL request for search via director & add log for search

twiks-yp-okyan committed:
fix EqualsAndHashCode annotation for Film entity

Alex-Galvinauskas committed:
Merge pull request #25 from Alex-Galvinauskas/add-common-films

Добавлена Функциональность «Общие фильмы»

Alex-Galvinauskas committed:
Merge pull request #26 from Alex-Galvinauskas/add-search

Add search

Alex-Galvinauskas committed:
fix: Исправлена загрузка жанров фильмов из БД

1. В методе getAllFilms() упрощен SQL-запрос и добавлена дозагрузка жанров
   для каждого фильма через genreDbStorage.getGenresByFilmId()
2. Во всех методах, возвращающих списки фильмов, добавлена дозагрузка жанров:
   - getPopularFilms()
   - getFilmsViaSearchByName()
   - getFilmsViaSearchByDirector()
   - getFilmsByDirector()
   - getPopularFilms(count, genreId, year)
   - getCommonFilms()
   - getRecommendationsForUser()
   - getFilmsByIds()
3. Добавлен .distinct() в saveFilmGenres() для предотвращения дублирования жанров
4. Улучшено логирование для отладки загрузки жанров

Исправлена ошибка теста: "genres field must be have 1 genre: expected +0 to deeply equal 1"
Теперь фильмы корректно возвращают жанры, сохраненные в таблице film_genres.

Alex-Galvinauskas committed:
fix: исправить часть ошибок

Alex-Galvinauskas committed:
Merge pull request #30 from Alex-Galvinauskas/FixDevelop

Fix develop

Alex-Galvinauskas committed:
Update DirectorDbStorageTest.java

forsidenis committed:
правка тестов

Alex-Galvinauskas committed:
Merge branch 'FixDevelop' into add-most-populars

Alex-Galvinauskas committed:
Merge pull request #32 from forsidenis/add-most-populars

add-most-populars

Alex-Galvinauskas committed:
Revert "add-most-populars"

Stepanovmaxvl committed:
Сделаны правки после ревью неуспешных тестов

Alex-Galvinauskas committed:
Merge pull request #33 from Alex-Galvinauskas/revert-32-add-most-populars

Revert "add-most-populars"

Stepanovmaxvl committed:
Merge remote-tracking branch 'origin/add-reviews' into add-reviews

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/film/validation/FilmValidatorImpl.java
#	src/main/java/ru/yandex/practicum/filmorate/service/review/ReviewServiceImpl.java

Stepanovmaxvl committed:
Merge branch 'FixDevelop' into add-reviews

Stepanovmaxvl committed:
Исправление

Alex-Galvinauskas committed:
Merge remote-tracking branch 'origin/FixDevelop' into FixDevelop

Alex-Galvinauskas committed:
Merge pull request #34 from Alex-Galvinauskas/add-reviews

Add reviews

Alex-Galvinauskas committed:
Bug: откат до более менее рабочей

Alex-Galvinauskas committed:
Merge branch 'FixDevelop' of github.com:Alex-Galvinauskas/java-filmorate into FixDevelop

Stepanovmaxvl committed:
Исправления

Alex-Galvinauskas committed:
Merge pull request #35 from Alex-Galvinauskas/FixDevelop

Fix develop

Alex-Galvinauskas committed:
style: почистить код от комментариев

fix: правки кода (0)

fix: еще правки кода

fix: снова правки (8)

fix: снова правки кода

fix: правки кода (24)

Alex-Galvinauskas committed:
fix: правки кода (24)

Alex-Galvinauskas committed:
fix: снова правки кода

Alex-Galvinauskas committed:
fix: снова правки (8)

Alex-Galvinauskas committed:
fix: еще правки кода

Alex-Galvinauskas committed:
fix: правки кода (0)

Alex-Galvinauskas committed:
style: почистить код от комментариев

Alex-Galvinauskas committed:
Merge branch 'develop' of github.com:Alex-Galvinauskas/java-filmorate into develop

Alex-Galvinauskas committed:
checkstyle: убрать импорты

Alex-Galvinauskas committed:
style: подчистил комментарии

Alex-Galvinauskas committed:
refactor: перенести поиск фильмов в отдельный сервис

fix: устранение circular dependency между FilmService и FilmSearchService

Alex-Galvinauskas committed:
feat: централизация логики валидации фильмов в FilmValidatorImpl

- Переименованы  FilmValidatorImpl и FilmValidatorRules в FilmValidatorService и FilmValidator
- Перемещены все методы валидации из FilmServiceImpl в FilmValidatorImpl:
  - validateMpa() - валидация MPA рейтинга
  - validateAndPrepareGenres() - валидация и подготовка жанров
  - validateAndPrepareDirectors() - валидация и подготовка режиссеров
  - validateCommonFilmsParams() - валидация параметров общих фильмов
  - validatePopularFilmsParams() - валидация параметров популярных фильмов
  - validateSearchParams() - валидация поисковых параметров
- Расширен интерфейс FilmValidator новыми методами валидации
- Обновлен FilmServiceImpl для использования методов валидатора
- Сохранена вся существующая бизнес-логика без изменений

maks-avilow commented on src/main/java/ru/yandex/practicum/filmorate/managment/db/FilmDbStorage.java:
```diff
@@ -69,9 +73,14 @@ public List<Film> getAllFilms() {
-                .peek(film -> film.setGenres(genreDbStorage.getGenresByFilmId(film.getId())))
-                .collect(Collectors.toList());
+        for (Film film : films) {
+            film.setGenres(genreDbStorage.getGenresByFilmId(film.getId()));
```
В классе FilmDbStorage в методах:
getAllFilms()
getFilmById()
getPopularFilms()
getFilmsViaSearchByName()
getFilmsViaSearchByDirector()
getFilmsByIds()
getRecommendationsForUser()
getFilmsByDirector()
и других
Происходит отдельный SQL-запрос для каждого фильма:
```
film.setGenres(genreDbStorage.getGenresByFilmId(film.getId()));
film.setDirectors(directorDbStorage.getDirectorsByFilmId(film.getId()));
film.setLikes(getLikesByFilmId(film.getId()));
```
Пример N+1:
Если getAllFilms() возвращает 100 фильмов → будет:

1 запрос на сами фильмы
+100 запросов на жанры
+100 запросов на режиссёров
+100 запросов на лайки
в итоге 301 запрос

Это классическая N+1 проблема. Это очень распространенная проблема, о которой вас обязательно спросят на собеседовании и как вы ее решали. Советую  подробно изучить эту тему и варианты решения:) ну и заодно исправить то о чем я написал

лучше бы загружать связанные сущности одним запросом с JOIN, а затем группировать. В getAllFilms() сделать JOIN с film_genres и genres
Получить все строки вида (film_id, film_data, genre_id, genre_name)
Сгруппировать по film_id - собрать Film с List<Genre>
То же - для режиссёров и лайков

Alex-Galvinauskas committed:
refactor: вынос логики для лайков и популярных фильмов в сервисы

- Создан LikeEventService для обработки событий лайков.
- Создан PopularFilmsService для работы с популярными фильмами.
- Обновлен FilmServiceImpl:
   - Удален приватный метод recordLikeEvent
   - Упрощен метод getPopularFilms (делегирование сервису)
   - Добавлены зависимости на новые сервисы

maks-avilow commented on src/main/java/ru/yandex/practicum/filmorate/managment/db/FilmDbStorage.java:
```diff
@@ -159,16 +173,79 @@ public List<Film> getPopularFilms(int count) {
-        return films.stream()
-                .peek(film -> film.setGenres(genreDbStorage.getGenresByFilmId(film.getId())))
-                .collect(Collectors.toList());
+        try {
```
метод  public List<Film> getPopularFilms(int count) {нигде не используется. что нарушает принцип yagni 
прошу его или использовать или удалить)

maks-avilow commented on src/main/java/ru/yandex/practicum/filmorate/managment/db/FilmDbStorage.java:
```diff
@@ -215,16 +295,332 @@ public Film mapRow(ResultSet rs, int rowNum) throws SQLException {
     public boolean existsFilmByNameAndReleaseYearExcludingId(String name, Integer releaseYear,
                                                              Long excludedId) {
-        String sql = "SELECT COUNT(*) FROM films WHERE LOWER(name) = LOWER(?) AND YEAR(release_date) = ? AND id != ?";
+        String sql = """
```
Здесь вложенный подзапрос в подзапросе
Внешний: NOT EXISTS (SELECT 1 FROM film_directors ... WHERE ... IN (...))
Внутренний: SELECT fd2.director_id FROM film_directors ...
Это нарушает наше требование к работе которое звучит как: "не допускаются запросы с двумя вложенными подзапросами"

прошу переписать через JOIN

Alex-Galvinauskas replied:
Спасибо за комментарии!

maks-avilow reviewed (changes requested):
Привет! Прежде всего хочу поблагодарить за проделанную большую работу:)
В ней очень много плюсов для данного этапа обучения и выполнено большинство требований. Что мне понравилось коротко:
-чистая архитектура слоёв
-чёткое разделение: Storage - Service - Controller . Нет логики в контроллерах, нет SQL в сервисах
- использование JdbcTemplate корректно
- все запросы параметризованы - нет SQL-инъекций. Хорошо обрабатываются KeyHolder, batchUpdate
-транзакционность
-аннотация @Transactional на уровне репозиториев - хотя чаще её ставят на сервисы, но в данном стеке (Spring Boot + JDBC) и так допустимо
-работа с коллекциями и Optional
- используется Optional, корректная обработка пустых результатов, исключения NotFoundException  соответствует REST- и clean code-практикам
-логирование: уместные log.debug/info/warn — помогают в диагностике без засорения
-валидация данных
-проверки вроде if (director.getName() == null || ...)  на месте
-инфраструктурная гибкость
-есть как in-memory, так и DB-реализации хранилищ (@Primary), что удобно для тестирования
-мапперы и DTO
-отделение DirectorDTO от Director хорошая практика
-безопасное формирование IN
Тем не менее у вас есть проблема n+1, оставил развернутый комментарий и прошу доработать)

Alex-Galvinauskas committed:
refactor: создать сервис лайков

Alex-Galvinauskas committed:
fix: убраны вложенные подзапросы

В методе existsFilmByNameAndReleaseYearExcludingId класса FilmDbStorage была
обнаружена проблема: использовался подзапрос внутри подзапроса, что нарушает
требования к кодовой базе.

Изменения:
- Переписан SQL-запрос для проверки существования фильма с таким же названием
  и годом выпуска, исключая текущий фильм
- Вместо вложенного подзапроса IN (SELECT ...) используется JOIN для соединения
  таблицы film_directors самой с собой
- Теперь запрос имеет только один уровень вложенности (NOT EXISTS), что
  соответствует требованиям
- Функциональность метода сохранена полностью"

Alex-Galvinauskas committed:
fix: решена проблема N+1 при загрузке связанных данных фильмов

Основные изменения:
1. Вынесен общий метод enrichFilmsWithRelatedData()
   Этот метод заменяет все отдельные вызовы загрузки связанных данных и решает проблему N+1.

2. Проблема N+1 решена за счет:
   - Загрузки жанров для всех фильмов одним запросом вместо 100 запросов для 100 фильмов
   - Загрузки режиссеров для всех фильмов одним запросом вместо 100 запросов
   - Загрузки лайков для всех фильмов одним запросом вместо 100 запросов

3. Вместо 301 запроса для 100 фильмов теперь будет всего 4 запроса:
   - Запрос на сами фильмы
   - Один запрос на все жанры для этих фильмов
   - Один запрос на всех режиссеров для этих фильмов
   - Один запрос на все лайки для этих фильмов

4. Все методы теперь используют единый подход:
   getAllFilms(), getFilmById(), getPopularFilms(), getFilmsViaSearchByName(),
   getFilmsViaSearchByDirector(), getFilmsByIds(), getRecommendationsForUser(),
   getFilmsByDirector(), getCommonFilms() - все они вызывают enrichFilmsWithRelatedData(),
   который загружает связанные данные оптимальным способом.

5. Логика не изменена:
   Классы FilmDTO.java и Film.java не изменялись, так как они не связаны с проблемой N+1.
   Вся оптимизация выполнена только в FilmDbStorage.java.

Новые методы:
- enrichFilmsWithRelatedData() - основной метод для загрузки связанных данных
- loadGenresForFilms() - загружает жанры для списка фильмов одним запросом
- loadDirectorsForFilms() - загружает режиссеров для списка фильмов одним запросом
- loadLikesForFilms() - загружает лайки для списка фильмов одним запросом

Alex-Galvinauskas committed:
Merge pull request #37 from Alex-Galvinauskas/Refactor-1

Refactor 1

Alex-Galvinauskas committed:
test: fix

Alex-Galvinauskas committed:
chekstyle: удалить неиспользуемые импорты

maks-avilow reviewed (approved):
зачет!
